File: card_generator.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Card Generator - Python Implementation
Based on the algorithm from https://gen.caicode.org/

This script generates virtual card numbers with expiry dates and CVV codes
using various input patterns and the Luhn algorithm for validation.
"""

import random
import re
from datetime import datetime
from typing import Dict, List, Tuple


class CardGenerator:
    """
    Card Generator class that implements the card generation algorithm
    with Luhn checksum validation.
    """

    def __init__(self):
        """Initialize the card generator."""
        self.current_year = datetime.now().year

    @staticmethod
    def luhn_checksum(card_number: str) -> int:
        """
        Calculate the Luhn checksum digit for a card number.

        Args:
            card_number: Card number string (may contain non-digits)

        Returns:
            The Luhn checksum digit (0-9)
        """
        # Remove non-digits
        digits = [int(d) for d in card_number if d.isdigit()]
        total = 0
        is_even = True

        # Process from right to left
        for digit in reversed(digits):
            if is_even:
                digit *= 2
                if digit > 9:
                    digit -= 9
            total += digit
            is_even = not is_even

        return (10 - (total % 10)) % 10

    @staticmethod
    def is_valid_luhn(card_number: str) -> bool:
        """
        Validate a card number using the Luhn algorithm.

        Args:
            card_number: Card number string to validate

        Returns:
            True if valid, False otherwise
        """
        # Remove non-digits
        digits = [int(d) for d in card_number if d.isdigit()]
        total = 0
        is_even = False

        # Process from right to left
        for digit in reversed(digits):
            if is_even:
                digit *= 2
                if digit > 9:
                    digit -= 9
            total += digit
            is_even = not is_even

        return total % 10 == 0

    @staticmethod
    def random_digit() -> str:
        """Generate a random digit (0-9)."""
        return str(random.randint(0, 9))

    @staticmethod
    def random_month() -> str:
        """Generate a random month (01-12)."""
        return str(random.randint(1, 12)).zfill(2)

    def random_year(self) -> str:
        """Generate a random year (current year + 1 to 5 years)."""
        return str(self.current_year + random.randint(1, 5))

    @staticmethod
    def random_cvv() -> str:
        """Generate a random CVV (100-999)."""
        return str(random.randint(100, 999))

    @staticmethod
    def parse_input_pattern(input_str: str) -> List[str]:
        """
        Parse the input pattern and split by various separators.

        Args:
            input_str: Input pattern string

        Returns:
            List of parsed parts
        """
        input_str = input_str.strip()

        # Determine separator and split
        if '|' in input_str:
            parts = input_str.split('|')
        elif '/' in input_str:
            parts = input_str.split('/')
        elif '-' in input_str:
            parts = input_str.split('-')
        elif ':' in input_str:
            parts = input_str.split(':')
        elif ',' in input_str:
            parts = input_str.split(',')
        elif ' ' in input_str:
            parts = re.split(r'\s+', input_str)
        else:
            parts = [input_str]

        # Clean up parts
        return [part.strip() for part in parts if part.strip()]

    def detect_pattern(self, input_str: str) -> str:
        """
        Detect and describe the input pattern type.

        Args:
            input_str: Input pattern string

        Returns:
            Description of the detected pattern
        """
        if not input_str:
            return ""

        parts = self.parse_input_pattern(input_str)
        mode_text = ""

        if len(parts) == 1:
            pattern = parts[0]

            if re.match(r'^\d{6}$', pattern):
                mode_text = "🎯 BIN Code Mode (6-digit bank identifier)"
            elif 'x' in pattern.lower():
                mode_text = "🎯 X Pattern Generation (x/X = random digits)"
            elif '*' in pattern:
                mode_text = "🎯 Asterisk Pattern (* = random digits)"
            elif '#' in pattern:
                mode_text = "🎯 Hash Pattern (# = random digits)"
            elif re.match(r'^\d+$', pattern) and len(pattern) > 6:
                mode_text = "🎯 Incomplete Card (auto-complete to 16 digits)"
            else:
                mode_text = "🎯 Custom Card Format"
        elif len(parts) == 2:
            mode_text = "🎯 Card + Date Mode"
        elif len(parts) == 3:
            mode_text = "🎯 Three-Part Mode (Card|MM|YY)"
        elif len(parts) == 4:
            random_keywords = ['(m)', '(y)', '(cvv)', 'mm', 'yy', 'cvv', 'rnd', 'random']
            has_random = any(part.lower() in random_keywords for part in parts)

            if has_random:
                mode_text = "🎯 Smart Random Mode"
            else:
                mode_text = "🎯 Fully Custom Mode"
        else:
            mode_text = "🎯 Multi-Part Custom Format"

        # Detect separator type
        if '|' in input_str:
            mode_text += " (Pipe Separated)"
        elif '/' in input_str:
            mode_text += " (Slash Separated)"
        elif '-' in input_str:
            mode_text += " (Dash Separated)"
        elif ':' in input_str:
            mode_text += " (Colon Separated)"
        elif ',' in input_str:
            mode_text += " (Comma Separated)"
        elif ' ' in input_str:
            mode_text += " (Space Separated)"

        return mode_text

    def generate_card_by_pattern(self, pattern: str) -> Dict[str, str]:
        """
        Generate a card based on the input pattern.

        Args:
            pattern: Input pattern string

        Returns:
            Dictionary with cardNumber, month, year, and cvv
        """
        parts = self.parse_input_pattern(pattern)
        card_number = ""
        month = ""
        year = ""
        cvv = ""

        if len(parts) == 1:
            input_part = parts[0]

            # BIN Code Mode (6-digit)
            if re.match(r'^\d{6}$', input_part):
                card_number = input_part
                for _ in range(9):
                    card_number += self.random_digit()
                card_number += str(self.luhn_checksum(card_number))

            # X Pattern
            elif 'x' in input_part.lower():
                card_number = re.sub(r'[xX]', lambda m: self.random_digit(), input_part)
                if len(card_number) == 15:
                    card_number += str(self.luhn_checksum(card_number))
                elif len(card_number) == 16:
                    card_number = card_number[:-1] + str(self.luhn_checksum(card_number[:-1]))

            # Asterisk Pattern
            elif '*' in input_part:
                card_number = input_part.replace('*', lambda: self.random_digit())
                # Use re.sub for proper replacement
                card_number = re.sub(r'\*', lambda m: self.random_digit(), input_part)
                while len(card_number) < 15:
                    card_number += self.random_digit()
                card_number += str(self.luhn_checksum(card_number))

            # Hash Pattern
            elif '#' in input_part:
                card_number = re.sub(r'#', lambda m: self.random_digit(), input_part)
                while len(card_number) < 15:
                    card_number += self.random_digit()
                card_number += str(self.luhn_checksum(card_number))

            # Pure digits - incomplete card
            elif re.match(r'^\d+$', input_part):
                card_number = input_part
                while len(card_number) < 15:
                    card_number += self.random_digit()
                card_number += str(self.luhn_checksum(card_number))

            # Random month, year, cvv
            month = self.random_month()
            year = self.random_year()[-2:]
            cvv = self.random_cvv()

        elif len(parts) >= 2:
            # Process card number part
            card_number = parts[0]

            if 'x' in card_number.lower():
                card_number = re.sub(r'[xX]', lambda m: self.random_digit(), card_number)
            elif '*' in card_number:
                card_number = re.sub(r'\*', lambda m: self.random_digit(), card_number)
            elif '#' in card_number:
                card_number = re.sub(r'#', lambda m: self.random_digit(), card_number)

            # Process month (part 1)
            if len(parts) > 1:
                month_part = parts[1].lower()
                if month_part in ['(m)', 'mm', 'rnd', 'random']:
                    month = self.random_month()
                else:
                    month = parts[1].zfill(2)

            # Process year (part 2)
            if len(parts) > 2:
                year_part = parts[2].lower()
                if year_part in ['(y)', 'yy', 'rnd', 'random']:
                    year = self.random_year()[-2:]
                else:
                    year_val = parts[2]
                    if len(year_val) == 4:
                        year = year_val[-2:]
                    else:
                        year = year_val.zfill(2)

            # Process CVV (part 3)
            if len(parts) > 3:
                cvv_part = parts[3].lower()
                if cvv_part in ['(cvv)', 'cvv', 'rnd', 'random']:
                    cvv = self.random_cvv()
                else:
                    cvv = parts[3].zfill(3)

            # Special handling for 2-part input
            if len(parts) == 2:
                second_part = parts[1]
                # MMYY format
                if len(second_part) == 4 and second_part.isdigit():
                    month = second_part[:2]
                    year = second_part[2:4]
                # MMYYCVV format (6 digits)
                elif len(second_part) == 6 and second_part.isdigit():
                    month = second_part[:2]
                    year = second_part[2:4]
                    cvv = second_part[4:6]

            # Complete card number to 16 digits
            while len(card_number) < 15:
                card_number += self.random_digit()

            if len(card_number) == 15:
                card_number += str(self.luhn_checksum(card_number))

        # Fill in any missing parts with random values
        if not month:
            month = self.random_month()
        if not year:
            year = self.random_year()[-2:]
        if not cvv:
            cvv = self.random_cvv()

        # Ensure proper formatting
        month = month.zfill(2)
        year = year.zfill(2)
        cvv = cvv.zfill(3)

        return {
            'cardNumber': card_number,
            'month': month,
            'year': year,
            'cvv': cvv
        }

    def generate_cards(self, pattern: str, count: int = 10) -> List[Dict[str, str]]:
        """
        Generate multiple cards based on the pattern.

        Args:
            pattern: Input pattern string
            count: Number of cards to generate (default: 10)

        Returns:
            List of card dictionaries
        """
        if not pattern.strip():
            raise ValueError("Pattern cannot be empty!")

        results = []
        attempts = 0
        max_attempts = count * 10  # Prevent infinite loop

        while len(results) < count and attempts < max_attempts:
            attempts += 1
            card = self.generate_card_by_pattern(pattern)

            # Validate with Luhn algorithm
            if self.is_valid_luhn(card['cardNumber']):
                results.append(card)
            else:
                print(f"⚠️  Invalid card generated (Luhn check failed): {card['cardNumber']}")

        return results

    @staticmethod
    def format_card_number(card_number: str) -> str:
        """
        Format card number with spaces every 4 digits.

        Args:
            card_number: Card number string

        Returns:
            Formatted card number
        """
        return re.sub(r'(\d{4})(?=\d)', r'\1 ', card_number)

    def format_card_output(self, card: Dict[str, str]) -> str:
        """
        Format card information for output.

        Args:
            card: Card dictionary

        Returns:
            Formatted string: Card|MM|YY|CVV
        """
        return f"{card['cardNumber']}|{card['month']}|{card['year']}|{card['cvv']}"


def main():
    """Main function to demonstrate the card generator."""
    generator = CardGenerator()

    # Example patterns
    examples = [
        "434256",
        "4342xxxxxxxx",
        "4342****",
        "434256|12|25|123",
        "434256/12/25/123",
        "434256-12-25-123",
        "434256|(m)|(y)|(cvv)",
        "434256|mm|yy|cvv",
        "434256|1225",
        "434256:rnd:rnd:rnd",
    ]

    print("=" * 80)
    print("Card Generator - Python Implementation")
    print("Based on https://gen.caicode.org/")
    print("=" * 80)
    print()

    # Let user choose a pattern or enter custom
    print("Example patterns:")
    for i, example in enumerate(examples, 1):
        mode = generator.detect_pattern(example)
        print(f"{i}. {example:25s} - {mode}")

    print()
    user_input = input("Enter a pattern or number (1-10) to select, or press Enter for default '434256': ").strip()

    # Check if user entered a number to select from examples
    if user_input.isdigit() and 1 <= int(user_input) <= len(examples):
        pattern = examples[int(user_input) - 1]
        print(f"📌 Selected example #{user_input}: {pattern}")
    elif not user_input:
        pattern = "434256"
    else:
        pattern = user_input

    print()
    print(f"Using pattern: {pattern}")
    print(f"Detected mode: {generator.detect_pattern(pattern)}")
    print()

    # Generate cards
    cards = generator.generate_cards(pattern, count=10)

    print(f"\n{'=' * 80}")
    print(f"Generated {len(cards)} cards:")
    print(f"{'=' * 80}\n")

    for i, card in enumerate(cards, 1):
        formatted = generator.format_card_number(card['cardNumber'])
        output = generator.format_card_output(card)
        is_valid = "✅" if generator.is_valid_luhn(card['cardNumber']) else "❌"

        print(f"{i:2d}. {formatted} | {card['month']}/{card['year']} | CVV: {card['cvv']} {is_valid}")
        print(f"    Output: {output}")

    print(f"\n{'=' * 80}")
    print("All cards copied to clipboard format:")
    print(f"{'=' * 80}\n")

    for card in cards:
        print(generator.format_card_output(card))


if __name__ == "__main__":
    main()

File: src/utils/cardGenerator.ts
/**
 * Card Generator - TypeScript Implementation
 * Based on the algorithm from https://gen.caicode.org/
 * 
 * This utility generates virtual card numbers with expiry dates and CVV codes
 * using various input patterns and the Luhn algorithm for validation.
 */

export interface GeneratedCard {
  cardNumber: string;
  month: string;
  year: string;
  cvv: string;
}

export class CardGenerator {
  private currentYear: number;

  constructor() {
    this.currentYear = new Date().getFullYear();
  }

  /**
   * Calculate the Luhn checksum digit for a card number.
   */
  static luhnChecksum(cardNumber: string): number {
    // Remove non-digits
    const digits = cardNumber.replace(/\D/g, '').split('').map(Number);
    let total = 0;
    let isEven = true;

    // Process from right to left
    for (let i = digits.length - 1; i >= 0; i--) {
      let digit = digits[i];
      if (isEven) {
        digit *= 2;
        if (digit > 9) {
          digit -= 9;
        }
      }
      total += digit;
      isEven = !isEven;
    }

    return (10 - (total % 10)) % 10;
  }

  /**
   * Validate a card number using the Luhn algorithm.
   */
  static isValidLuhn(cardNumber: string): boolean {
    const digits = cardNumber.replace(/\D/g, '').split('').map(Number);
    let total = 0;
    let isEven = false;

    // Process from right to left
    for (let i = digits.length - 1; i >= 0; i--) {
      let digit = digits[i];
      if (isEven) {
        digit *= 2;
        if (digit > 9) {
          digit -= 9;
        }
      }
      total += digit;
      isEven = !isEven;
    }

    return total % 10 === 0;
  }

  /**
   * Generate a random digit (0-9).
   */
  private randomDigit(): string {
    return Math.floor(Math.random() * 10).toString();
  }

  /**
   * Generate a random month (01-12).
   */
  private randomMonth(): string {
    return Math.floor(Math.random() * 12 + 1).toString().padStart(2, '0');
  }

  /**
   * Generate a random year (current year + 1 to 5 years).
   */
  private randomYear(): string {
    return (this.currentYear + Math.floor(Math.random() * 5 + 1)).toString();
  }

  /**
   * Generate a random CVV (100-999).
   */
  private randomCvv(): string {
    return Math.floor(Math.random() * 900 + 100).toString();
  }

  /**
   * Parse the input pattern and split by various separators.
   */
  private parseInputPattern(inputStr: string): string[] {
    const trimmed = inputStr.trim();

    let parts: string[];
    if (trimmed.includes('|')) {
      parts = trimmed.split('|');
    } else if (trimmed.includes('/')) {
      parts = trimmed.split('/');
    } else if (trimmed.includes('-')) {
      parts = trimmed.split('-');
    } else if (trimmed.includes(':')) {
      parts = trimmed.split(':');
    } else if (trimmed.includes(',')) {
      parts = trimmed.split(',');
    } else if (trimmed.includes(' ')) {
      parts = trimmed.split(/\s+/);
    } else {
      parts = [trimmed];
    }

    return parts.map(p => p.trim()).filter(p => p.length > 0);
  }

  /**
   * Detect and describe the input pattern type.
   */
  detectPattern(inputStr: string): string {
    if (!inputStr) {
      return '';
    }

    const parts = this.parseInputPattern(inputStr);
    let modeText = '';

    if (parts.length === 1) {
      const pattern = parts[0];

      if (/^\d{6}$/.test(pattern)) {
        modeText = '🎯 BIN Code Mode (6-digit bank identifier)';
      } else if (/x/i.test(pattern)) {
        modeText = '🎯 X Pattern Generation (x/X = random digits)';
      } else if (pattern.includes('*')) {
        modeText = '🎯 Asterisk Pattern (* = random digits)';
      } else if (pattern.includes('#')) {
        modeText = '🎯 Hash Pattern (# = random digits)';
      } else if (/^\d+$/.test(pattern) && pattern.length > 6) {
        modeText = '🎯 Incomplete Card (auto-complete to 16 digits)';
      } else {
        modeText = '🎯 Custom Card Format';
      }
    } else if (parts.length === 2) {
      modeText = '🎯 Card + Date Mode';
    } else if (parts.length === 3) {
      modeText = '🎯 Three-Part Mode (Card|MM|YY)';
    } else if (parts.length === 4) {
      const randomKeywords = ['(m)', '(y)', '(cvv)', 'mm', 'yy', 'cvv', 'rnd', 'random'];
      const hasRandom = parts.some(part => randomKeywords.includes(part.toLowerCase()));

      if (hasRandom) {
        modeText = '🎯 Smart Random Mode';
      } else {
        modeText = '🎯 Fully Custom Mode';
      }
    } else {
      modeText = '🎯 Multi-Part Custom Format';
    }

    // Detect separator type
    if (inputStr.includes('|')) {
      modeText += ' (Pipe Separated)';
    } else if (inputStr.includes('/')) {
      modeText += ' (Slash Separated)';
    } else if (inputStr.includes('-')) {
      modeText += ' (Dash Separated)';
    } else if (inputStr.includes(':')) {
      modeText += ' (Colon Separated)';
    } else if (inputStr.includes(',')) {
      modeText += ' (Comma Separated)';
    } else if (inputStr.includes(' ')) {
      modeText += ' (Space Separated)';
    }

    return modeText;
  }

  /**
   * Generate a card based on the input pattern.
   */
  generateCardByPattern(pattern: string): GeneratedCard {
    const parts = this.parseInputPattern(pattern);
    let cardNumber = '';
    let month = '';
    let year = '';
    let cvv = '';

    if (parts.length === 1) {
      const inputPart = parts[0];

      // BIN Code Mode (6-digit)
      if (/^\d{6}$/.test(inputPart)) {
        cardNumber = inputPart;
        for (let i = 0; i < 9; i++) {
          cardNumber += this.randomDigit();
        }
        cardNumber += CardGenerator.luhnChecksum(cardNumber).toString();
      }
      // X Pattern
      else if (/x/i.test(inputPart)) {
        cardNumber = inputPart.replace(/[xX]/g, () => this.randomDigit());
        if (cardNumber.length === 15) {
          cardNumber += CardGenerator.luhnChecksum(cardNumber).toString();
        } else if (cardNumber.length === 16) {
          cardNumber = cardNumber.slice(0, -1) + CardGenerator.luhnChecksum(cardNumber.slice(0, -1)).toString();
        }
      }
      // Asterisk Pattern
      else if (inputPart.includes('*')) {
        cardNumber = inputPart.replace(/\*/g, () => this.randomDigit());
        while (cardNumber.length < 15) {
          cardNumber += this.randomDigit();
        }
        cardNumber += CardGenerator.luhnChecksum(cardNumber).toString();
      }
      // Hash Pattern
      else if (inputPart.includes('#')) {
        cardNumber = inputPart.replace(/#/g, () => this.randomDigit());
        while (cardNumber.length < 15) {
          cardNumber += this.randomDigit();
        }
        cardNumber += CardGenerator.luhnChecksum(cardNumber).toString();
      }
      // Pure digits - incomplete card
      else if (/^\d+$/.test(inputPart)) {
        cardNumber = inputPart;
        while (cardNumber.length < 15) {
          cardNumber += this.randomDigit();
        }
        cardNumber += CardGenerator.luhnChecksum(cardNumber).toString();
      }

      // Random month, year, cvv
      month = this.randomMonth();
      year = this.randomYear().slice(-2);
      cvv = this.randomCvv();
    } else if (parts.length >= 2) {
      // Process card number part
      cardNumber = parts[0];

      if (/x/i.test(cardNumber)) {
        cardNumber = cardNumber.replace(/[xX]/g, () => this.randomDigit());
      } else if (cardNumber.includes('*')) {
        cardNumber = cardNumber.replace(/\*/g, () => this.randomDigit());
      } else if (cardNumber.includes('#')) {
        cardNumber = cardNumber.replace(/#/g, () => this.randomDigit());
      }

      // Process month (part 1)
      if (parts.length > 1) {
        const monthPart = parts[1].toLowerCase();
        if (['(m)', 'mm', 'rnd', 'random'].includes(monthPart)) {
          month = this.randomMonth();
        } else {
          month = parts[1].padStart(2, '0');
        }
      }

      // Process year (part 2)
      if (parts.length > 2) {
        const yearPart = parts[2].toLowerCase();
        if (['(y)', 'yy', 'rnd', 'random'].includes(yearPart)) {
          year = this.randomYear().slice(-2);
        } else {
          const yearVal = parts[2];
          if (yearVal.length === 4) {
            year = yearVal.slice(-2);
          } else {
            year = yearVal.padStart(2, '0');
          }
        }
      }

      // Process CVV (part 3)
      if (parts.length > 3) {
        const cvvPart = parts[3].toLowerCase();
        if (['(cvv)', 'cvv', 'rnd', 'random'].includes(cvvPart)) {
          cvv = this.randomCvv();
        } else {
          cvv = parts[3].padStart(3, '0');
        }
      }

      // Special handling for 2-part input
      if (parts.length === 2) {
        const secondPart = parts[1];
        // MMYY format
        if (secondPart.length === 4 && /^\d+$/.test(secondPart)) {
          month = secondPart.slice(0, 2);
          year = secondPart.slice(2, 4);
        }
        // MMYYCVV format (6 digits)
        else if (secondPart.length === 6 && /^\d+$/.test(secondPart)) {
          month = secondPart.slice(0, 2);
          year = secondPart.slice(2, 4);
          cvv = secondPart.slice(4, 6);
        }
      }

      // Complete card number to 16 digits
      while (cardNumber.length < 15) {
        cardNumber += this.randomDigit();
      }

      if (cardNumber.length === 15) {
        cardNumber += CardGenerator.luhnChecksum(cardNumber).toString();
      }
    }

    // Fill in any missing parts with random values
    if (!month) {
      month = this.randomMonth();
    }
    if (!year) {
      year = this.randomYear().slice(-2);
    }
    if (!cvv) {
      cvv = this.randomCvv();
    }

    // Ensure proper formatting
    month = month.padStart(2, '0');
    year = year.padStart(2, '0');
    cvv = cvv.padStart(3, '0');

    return {
      cardNumber,
      month,
      year,
      cvv,
    };
  }

  /**
   * Generate multiple cards based on the pattern.
   */
  generateCards(pattern: string, count: number = 10): GeneratedCard[] {
    if (!pattern.trim()) {
      throw new Error('Pattern cannot be empty!');
    }

    const results: GeneratedCard[] = [];
    let attempts = 0;
    const maxAttempts = count * 10; // Prevent infinite loop

    while (results.length < count && attempts < maxAttempts) {
      attempts++;
      const card = this.generateCardByPattern(pattern);

      // Validate with Luhn algorithm
      if (CardGenerator.isValidLuhn(card.cardNumber)) {
        results.push(card);
      } else {
        console.warn(`⚠️  Invalid card generated (Luhn check failed): ${card.cardNumber}`);
      }
    }

    return results;
  }

  /**
   * Format card number with spaces every 4 digits.
   */
  static formatCardNumber(cardNumber: string): string {
    return cardNumber.replace(/(\d{4})(?=\d)/g, '$1 ');
  }

  /**
   * Format card information for output.
   */
  static formatCardOutput(card: GeneratedCard): string {
    return `${card.cardNumber}|${card.month}|${card.year}|${card.cvv}`;
  }
}

